python/sparkdl/graph/input.py

```diff
@@ -0,0 +1,232 @@
+# Copyright 2017 Databricks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from __future__ import absolute_import, division, print_function
+
+import tensorflow as tf
+from tensorflow.core.protobuf import meta_graph_pb2
+
+import sparkdl.graph.utils as tfx
+
+__all__ = ["TFInputGraph"]
```

Comment:
I think this whole file has much more complexity that it really needs to, given the aim of its content.

We do not need a builder pattern, based on looking at the other PRs. You also happen to expose these builder classes implicitly because you return them, but this is not transparent from `__all__`. This is a limitation of python that does not help for software engineering.

Once we drop that, all this code really needs to expose is one immutable data structure and six functions that are stateless. As we discussed, there is no need to create additional classes, static methods or other python features. Everything here can be done with regular functions. Here is what I think it should look like:

```python
TFInputGraph=namedtuple(["graph_def", "inputs", "outputs"])
"""
A frozen representation of a tensorflow graph, and some extra information to map spark columns to
the graph's input and output tensors. Users should not have to peek into its content, or make
any assumption about the content.

graph_def: a tensorflow GraphDef object
inputs: a dictionary of {string: string} where the key is XXX and the value is XXX
outputs: XXX
"""

def fromGraph(graph, sess, feed_names, fetch_names):
    """
    Builds an internal representation of a tensorflow graph, based on a tf.Graph object
    :param graph: XXX
    :param sess: XXX
    :param feed_names: XXX <- put the details here!
    :param fetch_names: XXX <- put the details here!
    :return: a TFInputGraph object
    """
    raise

... other public functions with full documentation

... all the other private functions. They can be lighter on documentation, but they should not require any sophisticated python features. No extra classes are required.
```

You should also take a look at named tuples for building them, but there you should absolutely not need any of the more advanced features for them (for the time being).
https://docs.python.org/2/library/collections.html#collections.namedtuple

This is a significant rewrite, but it is going to dramatically improve the readability and the correctness in the face of future changes.

Reply:
Using `classmethod` to create class instances is a well-accepted approach. It is commonly used in [CPython implementation of datetime](https://github.com/python/cpython/blob/3.6/Lib/datetime.py#L1459-L1497).

Classes created from `collections.namedtuple` are self-documenting, resembling Scala's case classes to some degree. 

```
>>> ABC = namedtuple('ABC', 'a, b, c', verbose=True)
class ABC(tuple):
    'ABC(a, b, c)'

    __slots__ = ()

    _fields = ('a', 'b', 'c')

    def __new__(_cls, a, b, c):
        'Create new instance of ABC(a, b, c)'
        return _tuple.__new__(_cls, (a, b, c))

    @classmethod
    def _make(cls, iterable, new=tuple.__new__, len=len):
        'Make a new ABC object from a sequence or iterable'
        result = new(cls, iterable)
        if len(result) != 3:
            raise TypeError('Expected 3 arguments, got %d' % len(result))
        return result

    def __repr__(self):
        'Return a nicely formatted representation string'
        return 'ABC(a=%r, b=%r, c=%r)' % self

    def _asdict(self):
        'Return a new OrderedDict which maps field names to their values'
        return OrderedDict(zip(self._fields, self))

    def _replace(_self, **kwds):
        'Return a new ABC object replacing specified fields with new values'
        result = _self._make(map(kwds.pop, ('a', 'b', 'c'), _self))
        if kwds:
            raise ValueError('Got unexpected field names: %r' % kwds.keys())
        return result

    def __getnewargs__(self):
        'Return self as a plain tuple.  Used by copy and pickle.'
        return tuple(self)

    __dict__ = _property(_asdict)

    def __getstate__(self):
        'Exclude the OrderedDict from pickling'
        pass

    a = _property(_itemgetter(0), doc='Alias for field number 0')

    b = _property(_itemgetter(1), doc='Alias for field number 1')

    c = _property(_itemgetter(2), doc='Alias for field number 2')
```

You can see that they provide a lot more functions than we actually need.
In addition,  since the class inherit from `tuple`, the following should be noted.
```
>>> ABC = namedtuple('ABC', 'a, b, c'); isinstance(ABC(a=1, b=2, c=3), tuple)
True
```